go/arrow/compute/exprs/exec.go

```diff
@@ -0,0 +1,631 @@
+	}
+
+	return out, nil
+}
+
+func executeScalarBatch(ctx context.Context, input compute.ExecBatch, exp expr.Expression, ext ExtensionIDSet) (compute.Datum, error) {
+	if !exp.IsScalar() {
+		return nil, fmt.Errorf("%w: ExecuteScalarExpression cannot execute non-scalar expressions",
+			arrow.ErrInvalid)
+	}
+
+	switch e := exp.(type) {
+	case expr.Literal:
+		return literalToDatum(compute.GetAllocator(ctx), e, ext)
+	case *expr.FieldReference:
+		return execFieldRef(ctx, e, input, ext)
+	case *expr.Cast:
+		if e.Input == nil {
+			return nil, fmt.Errorf("%w: cast without argument to cast", arrow.ErrInvalid)
+		}
+
+		arg, err := executeScalarBatch(ctx, input, e.Input, ext)
+		if err != nil {
+			return nil, err
+		}
+		defer arg.Release()
+
+		dt, _, err := FromSubstraitType(e.Type, ext)
+		if err != nil {
+			return nil, fmt.Errorf("%w: could not determine type for cast", err)
+		}
+
+		var opts *compute.CastOptions
+		switch e.FailureBehavior {
+		case types.BehaviorThrowException:
+			opts = compute.SafeCastOptions(dt)
+		case types.BehaviorUnspecified:
+			opts = compute.UnsafeCastOptions(dt)
+		case types.BehaviorReturnNil:
+			return nil, fmt.Errorf("%w: cast behavior return nil", arrow.ErrNotImplemented)
+		}
+		return compute.CastDatum(ctx, arg, opts)
+	case *expr.ScalarFunction:
+		var (
+			err       error
+			allScalar = true
+			args      = make([]compute.Datum, e.NArgs())
+			argTypes  = make([]arrow.DataType, e.NArgs())
+		)
+		for i := 0; i < e.NArgs(); i++ {
+			switch v := e.Arg(i).(type) {
+			case types.Enum:
+				args[i] = compute.NewDatum(scalar.NewStringScalar(string(v)))
+			case expr.Expression:
+				args[i], err = executeScalarBatch(ctx, input, v, ext)
+				if err != nil {
+					return nil, err
+				}
+				defer args[i].Release()
+
+				if args[i].Kind() != compute.KindScalar {
+					allScalar = false
+				}
+			default:
+				return nil, arrow.ErrNotImplemented
+			}
+
+			argTypes[i] = args[i].(compute.ArrayLikeDatum).Type()
+		}
+
+		_, conv, ok := ext.DecodeFunction(e.FuncRef())
+		if !ok {
+			return nil, arrow.ErrNotImplemented
+		}
+
+		fname, opts, err := conv(e)
+		if err != nil {
+			return nil, err
+		}
+
+		ectx := compute.GetExecCtx(ctx)
+		fn, ok := ectx.Registry.GetFunction(fname)
+		if !ok {
+			return nil, arrow.ErrInvalid
+		}
+
+		if fn.Kind() != compute.FuncScalar {
+			return nil, arrow.ErrInvalid
+		}
+
+		k, err := fn.DispatchBest(argTypes...)
+		if err != nil {
+			return nil, err
+		}
+
+		var newArgs []compute.Datum
+		// cast arguments if necessary
+		for i, arg := range args {
+			if !arrow.TypeEqual(argTypes[i], arg.(compute.ArrayLikeDatum).Type()) {
+				if newArgs == nil {
+					newArgs = make([]compute.Datum, len(args))
+					copy(newArgs, args)
+				}
+				newArgs[i], err = compute.CastDatum(ctx, arg, compute.SafeCastOptions(argTypes[i]))
+				if err != nil {
+					return nil, err
+				}
+				defer newArgs[i].Release()
+			}
+		}
+		if newArgs != nil {
+			args = newArgs
+		}
```

Comment:
Technically this kind of implicit casting shouldn't be needed.  Substrait plans should always represent casts with explicit casts.

Reply:
Good point! this was carried over from my original work where I was evaluating general expressions before I converted this to evaluating substrait. I'll remove this to simplify it